src/Workspaces/Core/Portable/Workspace/Workspace.cs

```diff
@@ -1181,7 +1181,9 @@ private void CheckAllowedSolutionChanges(SolutionChanges solutionChanges)
         private void CheckAllowedProjectChanges(ProjectChanges projectChanges)
         {
             if (projectChanges.OldProject.CompilationOptions != projectChanges.NewProject.CompilationOptions
-                && !this.CanApplyChange(ApplyChangesKind.ChangeCompilationOptions))
+                && !this.CanApplyChange(ApplyChangesKind.ChangeCompilationOptions)
+                && !this.CanApplyCompilationOptionChange(
+                    projectChanges.OldProject.CompilationOptions, projectChanges.NewProject.CompilationOptions, projectChanges.NewProject))
             {
```

Comment:
Here you can see that if CanApplyChange is false but CanApplyCompilationOptionChange returns true, this check passes. Therefore VisualStudioWorkspace returning true for  CanApplyChange(ApplyChangesKind.ChangeParseOptions) CanApplyChange(ApplyChangesKind.ChangeCompilationOptions) must be wrong.

Reply:
@jcouv It looks like you wrote this. Could you clarify what the intent was here? Thanks

Reply:
I mean the equivalent code for parse options below.

Reply:
Hum, this does seem wrong. I would think that both checks should be:
```
if (<options changed>&& (!CanApplyChange() || !CanApplyXyzOptionChange()) ...
```
I'll let @jasonmalinowski confirm, since I'm not very familiar with workspaces.

---
In reply to: [178513336](https://github.com/dotnet/roslyn/pull/25875#discussion_r178513336) [](ancestors = 178513336)

Reply:
> since I'm not very familiar with workspaces

but you created this API 😄 

I'll mention the two options again. This API could be interpreted in 2 ways:
- `CanApplyChange(ApplyChangesKind.ChangeParseOptions)` returning true means the workspace is able to change *some* parse options - you don't know whether it can handle the option you want until you additionally call `CanApplyParseOptionChange`. So in this scenario, `CanApplyChange` really basically tells you nothing about what you can actually do. You might as well ignore it. This *might have been* the original intent (but implemented incorrectly).
-  `CanApplyChange(ApplyChangesKind.ChangeParseOptions)` returning true means the workspace supports changing parse options, period. Any kind of change, and you can count on that. Now it may also support changing some specific options even if this returns false. If you check `CanApplyParseOptionChange`, you might find out that is the case. This is the existing implementation.

@jcouv I see you're suggesting we should really have the first option. I think both of them make sense, but if we go with first, then the default value of `CanApplyCompilationChange` being false is wrong and would be a breaking change. Basically, until now, if you get true from `CanApplyChange(ApplyChangesKind.ChangeCompilationOptions)` you know the workspace supports changing any compilation options, but if we implement this the way you're suggesting, the new meaning of this would be "no, this actually doesn't supports changing compilation options because CanApplyCompilationChange still always returns false". If we go with number 1, the default implementation should return true. But there's also a problem that the existing implementation of `CanApplyParseChange` return false too. Meh :-/

Reply:
I'm leaning towards number 2 more and more because `CanApplyParseChange` and `CanApplyCompilationChange` are only protected, not public. If we go with number 1, if you ask `CanApplyChange(ApplyChangesKind.ChangeParseOptions)` and get true it would tell you absolutely nothing and you'd have no way to get the real answer. With number 2, you'd get false and miss out on an opportunity to actually make the change (which is a bummer), but getting true would still be a valuable information that you can definitely rely on.

Reply:
Thanks for the clarification.
The first reading is the idea that there is a coarse check that gates a more precise check. The second one is that either reason (the workspace can support any change, or can support this specific change) allows you to move forward.
Option (2) seems fine. And given that it's what's implemented, we can leave as-is (maybe add a comment?).

Reply:
If we keep option 2, we need to remove cases for ApplyChangesKind.ChangeParseOptions and ApplyChangesKind.ChangeCompilationOptions from CanApplyChange in VisualStudioWorkspace. See above: https://github.com/dotnet/roslyn/pull/25875/files#r178472162

Reply:
The intent of the API is it's a coarse check followed by precise check, i.e. it's the first of the two bullet points. As you can see, the precise check didn't exist until @jcouv added the additional way to do so, as _nobody_ ever did the coarse check because nobody ever implemented this anywhere. :smile: Practically, I expect people are now doing the CanApplyChange(ApplyChangesKind.ChangeParseOptions) as the coarse check, which is why I would be most worried that changing that to false is going to break things.

> If we go with number 1, if you ask CanApplyChange(ApplyChangesKind.ChangeParseOptions) and get true it would tell you absolutely nothing and you'd have no way to get the real answer. 

The fix is just make the second method public. It shouldn't have been protected in the first place.

Reply:
Thanks. It would make sense if this API was indeed public (and the default value of the precise check is change to `true` so we don't break existing implementors who use the coarse check as the only one). But I don't think I'd want to add a public API here, so then is it ok to leave that for later? What do you suggest I do about this now, leave everything as is, since things were equally broken before?

Reply:
I would either do nothing and leave all the brokeness as is, or make the API public and do the right thing. [insert Yoda meme here]

Reply:
If I did do that work, do you think it would be an acceptable change to change the default value of CanApplyParseOptionChange to true? I think that's the right value because anyone who doesn't implement that and only implements CanApplyChange expects users to be able to change those options. In fact if I kept these as returning false and fixed the code checking this, it would be a breaking change.

Reply:
Actually, default behavior should probably be to call CanApplyChange(ApplyChangesKind.ChangeParseOptions). That way workspaces that said no also say no from the new API, workspaces that say yes get the default "anything goes" which is the correct representation, and overriders get the right thing anyways.

Reply:
OK, I was just thinking that's a little silly/redundant if the proper check is to always to CanApplyChange **&&** CanApplyParseOptionChange anyway, but that works too I suppose.